### PR TITLE
feat: All changes required to change library to use ApiException and gapic internals

### LINKED ISSUE #148: Better exception for AdminClient
The exception that needs handling from the autocloseable `try(AdminClient adminClient = AdminClient.create(adminClientSettings)` is `java.lang.Exception`, which is very generic.

Pub/Sub uses `java.io.IOException` for `TopicAdminClient` and `SubscriptionAdminClient` ([samples](https://github.com/googleapis/java-pubsub/tree/master/samples/snippets/src/main/java/pubsub)).

See original [comment](https://github.com/googleapis/java-pubsublite/pull/147#discussion_r444510854) by @kurtisvg.

### LINKED ISSUE #180: Consider converting internals to use GAPIC layer
The internals are currently written using GRPC. We should investigate whether we can remove some of our code by converting this to use GAPIC, but we should not pursue this unless we can identify which pieces of code we can remove beforehand.

### PR DESCRIPTION
**Pull request by dpcollins-google**
This converts the internals of the client library to use GAPIC generated clients, as well as the api surface to raise ApiException as opposed to grpc StatusException.

fixes #180

fixes #148 

This does not throw RESOURCE_EXHAUSTED at high partition counts with the following code when removing RESOURCE_EXHAUSTED from the list of retryable errors:

```java 
public class PublisherTest {
  public static void main(String[] args) throws Exception {
    Publisher publisher = Publisher.create(PublisherSettings.newBuilder()
        .setTopicPath(TopicPath.parse(<400 partition topic>)).build());
    publisher.startAsync().awaitRunning();
    publisher.publish(PubsubMessage.newBuilder().setData(ByteString.copyFromUtf8("abc")).build()).get();
    publisher.stopAsync().awaitTerminated();
  }
}
```

### CODE REVIEW

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/CheckedApiException.java
Comment (manuelmenzella-google) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/CheckedApiException.java:23]: Nit: can you add a comment on why we need this? IIUC, this just wraps an ApiException but makes it checked for visibility.

**Review by manuelmenzella-google — APPROVED** [review]

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/CheckedApiException.java
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/CheckedApiException.java:23]: We don't 'need' this. Its to help with internal correctness, so that we have to handle errors that might arise. IMO, having ApiException inherit from RuntimeException is a mistake, but its what we have to work with.

Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/CheckedApiException.java:28]: Done.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/Subscriber.java
Comment (anguillanneuf) [inline google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/Subscriber.java:22]: Add description.

**Review by anguillanneuf — CHANGES_REQUESTED** [review]
Thanks for making this PR. I didn't review any of the tests. Here are some issues I found:

1. This comment about setting stub on the publisher needs an update: 
   https://github.com/googleapis/java-pubsublite/blob/7b6bd0545e49d7c7066764776b2cd2501b73a22f/google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/PublisherBuilder.java#L56
1. I'm not sure if you intended `google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/testing/UnitTestExamples.java` to be there.
1. Sporadic class descriptions in `google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/`.
1. Changes in `pubsublite-beam-io/` and `pubsublite-kafka-shim/` should be their own PRs. 
1. Changes to `grpc-google-cloud-pubsublite-v1/` should come from synthtool pulling from the source. 
1. Update samples in a later PR.
1. Also please resolve the conflicts with main.

I haven't tried the code to see how they impact samples myself. I can do that next week.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/ApiBackgroundResource.java
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/ApiBackgroundResource.java:26]: Nack. `@throws` does this automatically in javadoc

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/AssignerImpl.java
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/wire/AssignerImpl.java:82]: ApiException is unchecked. I introduced CheckedApiException for internal usage because I believe this to be a mistake, but it does not need to be declared.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/Subscriber.java
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/Subscriber.java:22]: Done.

File: google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/TrivialProxyService.java
Comment (dpcollins-google) [reply google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/TrivialProxyService.java:26]: Done.

**Comment by dpcollins-google** [comment]
> Thanks for making this PR. I didn't review any of the tests. Here are some issues I found:

Replies to each point below.

1) Done.
2) Yes, since this needs to be accessed from other modules.
3) This is not a public API, thus it can be put off until later as the only users of these classes would be maintainers.
4) No, since they would not compile otherwise.
5) I've reverted these. They were unintentional.
6) No, we've discussed this previously. For the integration tests to pass as currently set up, samples changes need to be made in the same PR, since the blocking presubmit is the snapshot one.
7) Done.

**Review by anguillanneuf — APPROVED** [review]
3. It's in the public [documentation](https://googleapis.dev/java/google-cloud-pubsublite/latest/index.html), either we add more documentation there, or we choose to hide it. Can be in a later PR. 
6. Ok. Make sure to make a new release soon after this PR is merged then.

**Pull request merged**
